### PR TITLE
Preserve last error for patchpoint helpers

### LINKED ISSUE #75828: [libraries-pgo] Issues related to overlapped IO
Multiple failures related to overlapped IO in net7.0-windows-Release-x64-CoreCLR_checked-jitosr_stress_random.

Work item link: https://dev.azure.com/dnceng-public/public/_build/results?buildId=21469&view=ms.vss-test-web.build-test-results-tab&runId=432776&paneView=debug

---
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-d8e4b98572834b0b9b/System.Net.NetworkInformation.Functional.Tests/1/console.cc044db2.log?helixlogtype=result

```scala
Unhandled exception. System.InvalidOperationException: 'overlapped' has already been freed.
   at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolBoundHandleOverlapped.cs:line 49
   at System.Threading.PortableThreadPool.IOCompletionPoller.Callback.Invoke(Event e) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.IO.Windows.cs:line 265
   at System.Threading.ThreadPoolTypedWorkItemQueue`2.System.Threading.IThreadPoolWorkItem.Execute() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs:line 1134
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs:line 919
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.cs:line 77
```

---
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-d8e4b98572834b0b9b/System.Net.Requests.Tests/1/console.7662a692.log?helixlogtype=result

```scala
Unhandled exception. System.InvalidOperationException: Handle is not initialized.
   at System.Threading.Overlapped.GetOverlappedFromNative(NativeOverlapped* pNativeOverlapped) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Overlapped.cs:line 228
   at System.Threading.IOCompletionCallbackHelper.PerformSingleIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/IOCompletionCallbackHelper.cs:line 37
   at System.Threading.PortableThreadPool.IOCompletionPoller.Callback.Invoke(Event e) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.IO.Windows.cs:line 265
   at System.Threading.ThreadPoolTypedWorkItemQueue`2.System.Threading.IThreadPoolWorkItem.Execute() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs:line 1134
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs:line 919
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.cs:line 77
```

---
Also, there are several AVs inside coreclr.dll and uses of corrupted OBJECTREFs in the same scenario.

### PR DESCRIPTION
**Pull request by AndyAyersMS**
In stress modes (and eventually perhaps in normal uses) the jit may insert patchpoint helper calls in regions where last error is live. So the helpers need to preserve last error.

Because some invocations of the helpers may transition to OSR methods instead of returning, we can't use the normal macros for this.

Fixes #75828.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@jakobbotsch PTAL
cc @dotnet/jit-contrib @jkotas

Seems like EnC could hit this issue too?

File: src/coreclr/vm/jithelpers.cpp [hunk 1]
```diff
@@ -4916,6 +4916,9 @@ HCIMPLEND
 
 void JIT_Patchpoint(int* counter, int ilOffset)
 {
+    // BEGIN_PRESERVE_LAST_ERROR;
+    DWORD __dwLastError = ::GetLastError();
```
Comment (jkotas) [inline src/coreclr/vm/jithelpers.cpp:4920]:
```suggestion
    DWORD dwLastError = ::GetLastError();
```

File: src/coreclr/vm/jithelpers.cpp [hunk 2]
```diff
@@ -5205,6 +5220,9 @@ void JIT_Patchpoint(int* counter, int ilOffset)
 //
 void JIT_PartialCompilationPatchpoint(int ilOffset)
 {
+    // BEGIN_PRESERVE_LAST_ERROR;
+    DWORD __dwLastError = ::GetLastError();
```
Comment (jkotas) [inline src/coreclr/vm/jithelpers.cpp:5224]:
```suggestion
    DWORD dwLastError = ::GetLastError();
```

**Review by jkotas — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
OSX seems to have some networking issues:
```
fatal: unable to access 'https://github.com/dotnet/runtime/': The requested URL returned error: 429
##[warning]Git fetch failed with exit code 128, back off 8.881 seconds before retry.
```

**Comment by stephentoub** [comment]
Yeah, almost every PR today has failed most mac legs because of that. @MattGal is investigating.

**Comment by MattGal** [comment]
> Yeah, almost every PR today has failed most mac legs because of that. @MattGal is investigating.

The "normal" channels are being a bit slow but I have confirmed they're aware of the throttling problem and are trying to identify the cause presently.

**Review by jakobbotsch — APPROVED** [review]

**Comment by jkotas** [comment]
Is this .NET 7 backport candidate? (Intermittent data corruption caused by OSR.)

**Comment by AndyAyersMS** [comment]
> Is this .NET 7 backport candidate? (Intermittent data corruption caused by OSR.)

I don't know if we can hit this with normal OSR.

The failure above is from a stress mode that places patchpoints in randomly selected blocks where the stack is empty, whereas OSR normally will only place patchpoints at blocks that are the start or end of a loop.

I'll spend some time today trying to work up a case; seems like the "end of loop" patchpoint (which is the default placement) might be a place we could hit this normally.

**Comment by AndyAyersMS** [comment]
> I would expect that something like this can hit this reliably:

Yes, thanks. 

Repro is simple and deterministic; the failure happens in Tier0 code once it has invoked the patchpoint helper. Will add a test case.

**Comment by AndyAyersMS** [comment]
For Linxu arm64 NAOT: msbuild crashed during restore:
```
/__w/1/s/.dotnet/sdk/7.0.100-preview.7.22377.5/MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:/__w/1/s/artifacts/log/Release/ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:false /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=true /p:__ToolsetLocationOutputFile=/__w/1/s/artifacts/toolset/8.0.0-beta.22469.1.txt /t:__WriteToolsetLocation /warnaserror /__w/1/s/artifacts/toolset/restore.proj
/__w/1/s/eng/common/tools.sh: line 474:   535 Segmentation fault      (core dumped) "$_InitializeBuildTool" "$@"
Build failed with exit code 139. Check errors above.
Finishing: Build product
```

**Comment by AndyAyersMS** [comment]
@jeffschwMSFT @JulieLeeMSFT we should consider porting this to 7.0.

**Comment by jeffschwMSFT** [comment]
@AndyAyersMS in what cases (beyond stress modes) do we anticipate that this may occur?

**Comment by AndyAyersMS** [comment]
> @AndyAyersMS in what cases (beyond stress modes) do we anticipate that this may occur?

This PR has a simple test case that fails without stress -- the value of last error will not be preserved across loops.

**Comment by jkotas** [comment]
In a more real-world situation, the long-running for loop in the test can be some kind of interop marshalling code.

**Pull request merged**